### PR TITLE
Feat/Login-Flow

### PR DESCRIPTION
**Pull request by joaooliveiradev**
- Feat: Add Auth Context
- Feat: Add Mutation Hook 
- Feat: Add Modal of Login and Signup
- Feat: Add logic to login in application and logout
- Feat: Add logic to save the token and username on cookie/local storage
- Feat: Add Error Message component to display error on SingIn Modal
- Feat: Add an touched prop on input to display error only when user is typing on that determined input
- Feat: Add Yup and SignIn Schema to validate user typing on form
- Feat: Add an function to destroy cookies, this is when user logout and to use in the future
- Feat: Add an function to handle with errors return from API
- Feat: Add some path alias on tsconfig to handle with imports
- Feat: Add React Query to handle with request and Formik to handle with forms
- Feat: Add React Query Config on app and a config to use React Query on SSR
- Refactor: Change Header component to receive the user from cookies rather than prop

### CODE REVIEW

File: src/components/Form/SignInModal/SignIn/index.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  const formik = useFormik({
+    initialValues: initialValues,
+    validationSchema: signInSchema,
+    onSubmit: (values: SignInValues) => handleSubmit(values),
+  })
+
+  const handleSubmit = (values: SignInValues) => {
+    const newSignInValues = {
+      user: {
+        ...values,
```
Comment (vmarcosp) [inline src/components/Form/SignInModal/SignIn/index.tsx:40]: Porque spread ao invés de só atribuir values para user?

Comment (joaooliveiradev) [reply src/components/Form/SignInModal/SignIn/index.tsx:40]: Realmente os dois dão na mesma, deixei passar batido. I will fix this.

**Comment by vmarcosp** [comment]
Peguei também dois comportamentos estranhos:
1. Ao clicar na parte branca do modal, duas vezes ou as vezes uma, ele fecha o modal:

https://user-images.githubusercontent.com/20327229/199110467-538f8779-b3b5-46f5-bee2-a8595ab99450.mov


2. Ao clicar em "Sign up" ~ as vezes ~ o modal não troca por conta do erro no campo email:

https://user-images.githubusercontent.com/20327229/199110544-3b5e23f2-6267-4c4e-bd6f-df015d6776d2.mov

**Review by vmarcosp — CHANGES_REQUESTED** [review]

File: src/schemas/Signin.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { object, string } from 'yup'
```
[the remark this answers is not in the document]
Comment (joaooliveiradev) [reply src/schemas/Signin.ts:1]: Então, quando é usado em um lugar só, desacoplar em uma pasta só por causa disso começa a ser desacoplamento desnecessário. Porém, uma dos motivos do porque eu pensei nisso, foi o seguinte, ja cheguei a fazer paginas com muitos campos de formulário e isso acaba deixando o schema muito grande, juntando isso numa página com o componente em si, estilos, acaba ficando gigante para procurar infos e tal, então eu escolhi deixar desacoplado, porém, analisando agora o projeto, os formulários não parecem tão grandes com bastante campos então realmente dá pra deixar tudo em um arquivo só. Vou resolver isso

**Comment by joaooliveiradev** [comment]
In this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/389c018a526ed37232c5cefefe4a8b83743079e4) i fix this request changes.

- The two strange behaviours on Modal.
- Move the Sign In schema on your own file.

**Comment by joaooliveiradev** [comment]
In this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/3827691512caf76f817f0fe3b4e42a8a7dd1ed9f) i fix this request changes.

- Feat: Make the hook useMe to get the user data rather than use local storage.
- Refactor: Remove Type from the DefaultErrorType name.
- Refactor: Put some fetcher configs inside of the your own file to let standart on all the application

File: src/hooks/queries/useMe.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+const UseMeResponseCodec = t.type({
+  user: UserTypeCodec
+})
+
+type UseMeResponse = t.TypeOf<typeof UseMeResponseCodec>
+
+export type UseMeOutput = t.OutputOf<typeof UseMeResponseCodec>
+
+export const useMe = async (): Promise<Either<DefaultErrorProps, UseMeOutput>> => {
```
Comment (vmarcosp) [inline src/hooks/queries/useMe.ts:15]: Não acha que isso aqui devia ser um hook ao invés de só uma função?

File: src/components/Header/index.tsx [hunk 4]
```diff
@@ -4,28 +4,59 @@ import Button from '@components/Button'
...
 import Image from 'next/image'
 import styled from 'styled-components'
-
-export type HeaderProps = {
-  isUserLoggedIn: boolean
-}
+import SignInModal from '@components/Form/SignInModal'
+import { useState } from 'react'
+import { useAuth } from '@context/auth'
+import { useMe, UseMeOutput } from '@hooks/queries/useMe'
```
Comment (vmarcosp) [inline src/components/Header/index.tsx:10]: Esse import poderia ser mais simples, do tipo `import { useMe, UseMeOutput } from "@hooks/queries"` ou apenas "@hooks"

File: src/components/Form/SignInModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import SignUp from '@components/Form/SignInModal/SignUp'
+import SignIn from '@components/Form/SignInModal/SignIn'
+import { useState } from 'react'
+import { Modal } from '@components/Modal'
```
Comment (vmarcosp) [inline src/components/Form/SignInModal/index.tsx:4]: Simplificar imports

**Review by vmarcosp — CHANGES_REQUESTED** [review]

File: src/hooks/queries/useMe.ts (hunk 3 above)
Comment (joaooliveiradev) [reply src/hooks/queries/useMe.ts]: I fix this in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/b3f897da8729940a69158281e93fdf538178df36)

File: src/components/Header/index.tsx (hunk 4 above)
Comment (joaooliveiradev) [reply src/components/Header/index.tsx]: I fix this in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/b3f897da8729940a69158281e93fdf538178df36)

File: src/components/Form/SignInModal/index.tsx (hunk 5 above)
Comment (joaooliveiradev) [reply src/components/Form/SignInModal/index.tsx]: I make a better way of import modules in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/c51d00664c6dae3875c8b8052f0a073cebda6135)

**Comment by joaooliveiradev** [comment]
In this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/a1323cee977f9525b9eb70ace7cb280c39fbb4a5) i fix that when the codec doens't match, i always returned an Unkown error, now i return a left of an DecodeError that contains the decodeErrors that i can debug if i want.

**Comment by joaooliveiradev** [comment]
In this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/3235ef0c6b1fd7a319b444f4b7ff3cb5a6964303) i change the errorMsg state from authContext to be a option of string.

File: src/components/ProfileName/ProfileName.stories.tsx [hunk 6]
```diff
@@ -1,5 +1,5 @@
 import { Meta, Story } from '@storybook/react'
-import ProfileName, { ProfileNameProps } from '@components/ProfileName'
+import { ProfileNameProps, ProfileName } from '@components/ProfileName'
```
Comment (vmarcosp) [inline src/components/ProfileName/ProfileName.stories.tsx:2]: Aqui também, deveria ser só "@components"

File: src/pages/_app.tsx [hunk 7]
```diff
@@ -7,8 +7,8 @@ import { useState } from 'react'
 import { Layout } from '@components/Layout'
 import type { AppProps } from 'next/app'
 import { ThemeProvider } from 'styled-components'
-import GlobalStyles from 'styles/global'
-import theme from 'styles/theme'
+import GlobalStyles from '@/styles/global'
```
Comment (vmarcosp) [inline src/pages/_app.tsx:10]: Não acha que poderia ser só @/styles?

**Review by vmarcosp — CHANGES_REQUESTED** [review]

File: src/components/Header/index.tsx [hunk 8]
```diff
@@ -4,28 +4,48 @@ import Button from '@components/Button'
...
 `
 
-export const Header = ({ isUserLoggedIn }: HeaderProps) => {
+export const Header = () => {
+  const [username, setUsername] = useState<string>('')
+  const [isModalOpen, setIsModalOpen] = useState<boolean>(false)
+  const { status, signOut } = useAuth()
+  useEffect(() => {
+    if (status == 'loggedIn') {
+      const storageUsername = localStorage.getItem('username')
```
[the remark this answers is not in the document]
Comment (vmarcosp) [reply src/components/Header/index.tsx]: Não esquece esse carinha @joaooliveiradev, trocar pra usar o useMe ao invés de armazenar no localStorage.

File: src/context/auth.tsx [hunk 9]
```diff
@@ -37,14 +35,14 @@ const defaultValueContext: ContextProps = {
...
-  errorMsg: '',
+  errorMsg: none,
 }
 
 const AuthContext = createContext(defaultValueContext)
 
 const AuthProvider = ({ children }: AuthContextProps) => {
   const [status, setStatus] = useState<Status>('idle')
-  const [errorMsg, setErrorMsg] = useState<string>('')
+  const [errorMsg, setErrorMsg] = useState<Option<string>>(none)
```
[the remark this answers is not in the document]
Comment (joaooliveiradev) [reply src/context/auth.tsx]: What you think about this [solution](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/04e83d09fa7f61376baee79b829db7d06c268224)

File: src/pages/_app.tsx [hunk 10]
```diff
@@ -7,8 +7,8 @@ import { useState } from 'react'
...
+import theme from '@/styles/theme'
```
[the remark this answers is not in the document]
Comment (joaooliveiradev) [reply src/pages/_app.tsx]: In this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/b536b1080a021cadfc4ff996ccb8a567cd4df3d9) i make an refactor that simplify more imports

**Comment by joaooliveiradev** [comment]
when i was doing some request changes i realize that the input color does not change your color to red when was an error. This behaviour had before, but i believe with some changes that i made to put error messages with touched, the behaviour stopped working, so i implement again in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/be64883274c94670fd358679dbe0bfc5f63d60f1)

File: src/context/auth.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import { signInMutation, type SignInResponseOutput } from './signInMutation'
+
+type Status = 'loggedOut' | 'loggedIn' | 'idle'
+
+type ContextProps = {
+  status: Status
+  signIn: (values: SignInInput) => void
+  isLoading: boolean
+  signOut: () => void
+  data: Either<DefaultError, SignInResponseOutput> | undefined
```
Comment (vmarcosp) [inline src/context/auth.tsx:22]: Porque usar `undefined` e não option do fp-ts?

**Review by vmarcosp — CHANGES_REQUESTED** [review]

File: src/components/Form/SignInModal/SignUp/index.tsx [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  Text,
+  ChangeFormBtn,
+} from '@/components/Form/SignInModal/styles'
+
+
+type SignUpProps = {
+  handleClick: (state: boolean) => void
+}
+
+export const SignUp = ({ handleClick }: SignUpProps) => {
```
[the remark this answers is not in the document]
Comment (joaooliveiradev) [reply src/components/Form/SignInModal/SignUp/index.tsx]: I fix in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/3d50255ed798300d8f33f719b980089a603f75e2)

File: src/components/Form/SignInModal/SignIn/index.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    .email('E-mail is not valid')
+    .required('Email field is required')
+    .lowercase(),
+  password: Yup.string()
+    .required('Password field is required')
+    .min(8, 'Minimum 8 characters required!')
+    .max(64, 'Maximum 64 characters required!'),
+})
+
+const SignIn = ({ handleClick }: SignInProps) => {
```
[the remark this answers is not in the document]
Comment (joaooliveiradev) [reply src/components/Form/SignInModal/SignIn/index.tsx]: I fix in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/3d50255ed798300d8f33f719b980089a603f75e2)

File: src/context/auth.tsx (hunk 11 above)
Comment (joaooliveiradev) [reply src/context/auth.tsx]: I fix in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/dfb7f6ff11ff59605c6f9b54956484203da980a4)

File: src/context/auth.tsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+  error: DefaultError | null
```
[the remark this answers is not in the document]
Comment (joaooliveiradev) [reply src/context/auth.tsx]: I fix in this [commit](https://github.com/joaooliveiradev/conduit-frontend/pull/35/commits/dfb7f6ff11ff59605c6f9b54956484203da980a4)

File: src/components/Form/SignInModal/SignIn/index.tsx [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+  })
+
+  const handleSubmit = (values: SignInValues) => {
+    const newSignInValues = {
+      user: values,
+    }
+    signIn(newSignInValues)
+  }
+
+  const dataError = pipe(
```
Comment (vmarcosp) [inline src/components/Form/SignInModal/SignIn/index.tsx:55]: `**` Não sei, não to curtindo muito esse código aqui. Vamos falar em call sobre ele.
Vou anotar com ** tudo que formos falar em call.

File: src/components/Form/SignInModal/SignIn/index.tsx [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    data,
+    fromNullable,
+    chain((data) => isSome(data) ? getLeft(data.value) : none)
+  )
+
+  const handleError = () => {
```
Comment (vmarcosp) [inline src/components/Form/SignInModal/SignIn/index.tsx:61]: `**`

File: src/components/Header/index.tsx [hunk 17]
```diff
@@ -1,31 +1,60 @@
...
   justify-content: space-between;
 `
 
-export const Header = ({ isUserLoggedIn }: HeaderProps) => {
+export const Header = () => {
+  const [isModalOpen, setIsModalOpen] = useState<boolean>(false)
+  const { status, signOut } = useAuth()
+  const { data } = useMe()
+
+  const username = pipe(
```
Comment (vmarcosp) [inline src/components/Header/index.tsx:29]: `**`

File: src/context/auth.tsx [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+const AuthContext = createContext(defaultValueContext)
+
+const AuthProvider = ({ children }: AuthContextProps) => {
+  const [status, setStatus] = useState<Status>('idle')
+
+  useEffect(() => {
+    const { 'conduit.token': accessToken } = parseCookies()
+    if (accessToken) return setStatus('loggedIn')
```
Comment (vmarcosp) [inline src/context/auth.tsx:48]: Porque tem `return` aqui?

File: src/hooks/queries/useMe.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+const UseMeResponseCodec = t.type({
+  user: UserTypeCodec,
+})
+
+type UseMeResponse = t.TypeOf<typeof UseMeResponseCodec>
+
+type UseMeOutput = t.OutputOf<typeof UseMeResponseCodec>
+
+const useMe = async () =>
```
Comment (vmarcosp) [inline src/hooks/queries/useMe.ts:16]: `**`

**Review by vmarcosp — CHANGES_REQUESTED** [review]

**Review by vmarcosp — APPROVED** [review]

**Pull request merged**